getopt/getopt.c

```diff
@@ -0,0 +1,713 @@
+        return -1;
+
+    getopt_param param = scan_param(optstring ? optstring : "");
+
+    if(optind >= argc)
+    {
+        if(!param.posix && !param.process_all)
+            swap_values(argc, (char**)argv, optind);
+        return -1;
+    }
+
+    optarg = 0;
+    
+    if(!nextchar) // we are in begin of argument... no previous iterations on this arg
+    {
+        int retval;
+        if(!detect_next_option(argc, argv, &param, &retval))
+            return retval;
+        nextchar = short_option_begin(argv[optind]);
+    }
+
+    return get_short_option(argc, argv, &param);
+}
+
+int getopt_long(int argc, char * const argv[],
+                const char *optstring,
+                const struct option *longopts, int *longindex)
+{
+    if(!argc || !argv)
+        return -1;
+
+    getopt_param param = scan_param(optstring ? optstring : "");
+
+    if(optind >= argc)
+    {
+        if(!param.posix && !param.process_all)
+            swap_values(argc, (char**)argv, optind);
+        return -1;
+    }
+
+    optarg = 0;
+    
+    if(!nextchar) // we are in begin of argument... no previous iterations on this arg
+    {
+        int retval;
+        if(!detect_next_option(argc, argv, &param, &retval))
+            return retval;
+
+        const char* str = argv[optind];
+        if(is_long_option(str))
+        {
+            int _longindex;
+            if(!longindex) // if no longindex specified - create own temp...
+                longindex = &_longindex;
+            return get_long_option(argc, argv, &param, longopts, longindex);
+        }
+        else if(is_short_option(str))
+        {
+            nextchar = short_option_begin(str);
+        }
+    }
+
+    return get_short_option(argc, argv, &param);
+}
+
+int getopt_long_only(int argc, char * const argv[],
+                     const char *optstring,
+                     const struct option *longopts, int *longindex)
+{
+    if(!argc || !argv)
+        return -1;
+
+    getopt_param param = scan_param(optstring ? optstring : "");
+
+    if(optind >= argc)
+    {
+        if(!param.posix && !param.process_all)
+            swap_values(argc, (char**)argv, optind);
+        return -1;
+    }
+
+    optarg = 0;
+    
+    if(!nextchar) // we are in begin of argument... no previous iterations on this arg
+    {
+        int retval;
+        if(!detect_next_option(argc, argv, &param, &retval))
+            return retval;
+
+        const char* str = argv[optind];
+        if(is_long_option(str))
+        {
+            int _longindex;
+            if(!longindex) // if no longindex specified - create own temp...
+                longindex = &_longindex;
+            return get_long_option(argc, argv, &param, longopts, longindex);
+        }
+        else if(is_short_option(str))
+        {
+            return get_long_short_option(argc, argv, &param, longopts, longindex);
+        }
+    }
+
+    return get_short_option(argc, argv, &param);
+}
+
+
+
+
```

Comment:
The formatting of this file does not match the rest of the code base.  There's also no license at the top of the code.  We need to know the origin of this code and what license it had.

If this code is only needed for Windows, we may want to move it to an OS specific directory.